File: tools/probes/keywords.py
# SPDX-License-Identifier: GPL-3.0-or-later
# Part of the Iskra-226 project: https://github.com/Ptr314/Iskra-226
"""Таблица ключевых слов BASIC 02 из самого интерпретатора.

На загрузочной дискете, начиная с 0x1840, лежат три вещи подряд:

  1. алфавитный список однословных ключевых слов, склеенных без разделителей;
  2. список сочетаний: байт «индекс первого слова + 0x60» и текст продолжения;
  3. массив перестановки: по байту на запись обоих списков — её токен.

Границы слов в первом списке машина знает по своему разбору, а нам их
приходится задать руками: список алфавитный, и разбивка проверяется тем, что
склейка обратно совпадает байт в байт. Разбор целиком — docs/format.md,
разд. 4.

    py tools/probes/keywords.py [ОБРАЗ]
"""

import os
import sys

# Список 1: алфавитный, 51 слово. Проверяется склейкой (см. ниже).
WORDS = (
    'ADD AND( BACKSPACE BIN( BOOL CLEAR COM CONVERT COPY DATA DBACKSPACE '
    'DEFFN DIM DSKIP END FOR GOSUB GOTO HEXPRINT IF INIT INPUT KEYIN LET '
    'LIMITS LIST LOAD MOVE NEXT ON OR( PACK( PRINT READ REM RENUMBER RES '
    'RETURN REWIND ROTATE RUN SAVE SCRATCH SELECT SKIP STOP TRACE '
    'UNPACK( VERIFY XOR( $GIO'
).split()

START = 0x1840          # начало списка 1
COMBO_END = 0x19B5      # за последним сочетанием
PERM = 0x19B6           # массив перестановки

DEFAULT = os.path.join('..', 'DSKCommander', 'docs', 'samples', 'Iskra226',
                       '345_dk_spb', 'disk1side0.dsk')


def decode(data):
    blob = ''.join(WORDS)
    got = data[START:START + len(blob)].decode('latin1')
    if got != blob:
        raise SystemExit('список слов не сходится: %r' % got[:60])

    # Сочетания: код продолжения — байт вне латинских прописных, они и
    # разделяют записи.
    tail = data[START + len(blob):COMBO_END]
    names = list(WORDS)
    combos = []
    i = 0
    while i < len(tail):
        code = tail[i]
        i += 1
        text = ''
        while i < len(tail) and not (tail[i] >= 0x60 and
                                     not 0x41 <= tail[i] <= 0x5A):
            text += chr(tail[i])
            i += 1
        combos.append((code - 0x60, text))
        names.append(names[code - 0x60] + ' ' + text)

    perm = data[PERM:PERM + len(names)]
    if len(perm) != len(names):
        raise SystemExit('перестановка короче списка')
    return names, combos, perm


def main():
    path = sys.argv[1] if len(sys.argv) > 1 else DEFAULT
    data = open(path, 'rb').read()
    names, combos, perm = decode(data)
    print('слов %d, сочетаний %d, перестановка %d байт'
          % (len(WORDS), len(combos), len(perm)))
    for name, token in sorted(zip(names, perm), key=lambda p: p[1]):
        print('%02X  %s' % (token, name))


if __name__ == '__main__':
    main()

File: src/host_common/keywords.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Mikhail Revzin <p3.141592653589793238462643@gmail.com>
// Part of the Iskra-226 project: https://github.com/Ptr314/Iskra-226
// Description: верхний регистр зоны 1 — ключевые слова Бейсика по одной клавише

#include "host_common/keywords.h"

#include <string>

#include "core/koi8.h"

namespace iskra {

namespace {

// Знаки клавиши в UTF-8 (нижний и верхний регистры) и слово, которое она
// вводит. Порядок — как на рис. 2.2, по зонам.
//
// Три надписи на рисунке обрезаны шириной клавиши и дописаны по смыслу:
// `PRINTUS` → `PRINTUSING`, `RENUMB` → `RENUMBER`, `BACKSP` → `BACKSPACE`.
// Одну прочитать не удалось, и клавиши у неё здесь нет: `Д`/`D`, надпись
// `RES` обрезана. Выдумывать её незачем — клавиша просто не работает.
//
// Где знак клавиши совпадает с арифметическим знаком цифрового блока
// (`+`, `-`, `*`, `/`), взят только второй знак клавиши: цифровой блок
// отдан функциям зоны 5, и по знаку их не различить.
struct Entry {
    const char * chars;
    const char * word;
};

const Entry TABLE[] = {
    // Зона 1, буквы
    { "ЙJ", "FOR" },
    { "ЦC", "¤GIO" },
    { "УU", "GOSUB" },
    { "КK", "GOSUB'" },
    { "ЕE", "GOTO" },
    { "НN", "HEX(" },
    { "ГG", "HEXPRINT" },
    { "Ш[", "IF" },
    { "Щ]", "INIT" },
    { "ЗZ", "INPUT" },
    { "ХH", "INT(" },
    { "ФF", "NEXT" },
    { "ЫY", "ON" },
    { "ВW", "OR(" },
    { "АA", "PACK(" },
    { "ПP", "PRINTUSING" },
    { "РR", "READ" },
    { "ОO", "REM" },
    { "ЛL", "RENUMBER" },
    { "ЖV", "RESTORE" },
    { "Э\\", "RETURN" },
    { "ЯQ", "RND(" },
    { "Ч^", "ROTATE" },
    { "СS", "SELECT" },
    { "МM", "SKIP" },
    { "ИI", "SGN(" },
    { "ТT", "STEP" },
    { "ЬX", "STOP" },
    { "БB", "STR(" },
    { "Ю@", "TAB(" },
    // Зона 1, цифры и знаки
    { "1!", "AND(" },
    { "2\"", "BACKSPACE" },
    { "3#", "BIN(" },
    { "4$", "BOOL" },
    { "5%", "COM" },
    { "6&", "CONVERT" },
    { "7'", "DATA" },
    { "8(", "DEFFN" },
    { "9)", "DEFFN'" },
    { "0",  "DIM" },
    { ":",  "KEYIN" },
    { "=",  "END" },
    { ".>", "REWIND" },
    { ",<", "TEN" },
    { "?",  "TRACE" },
    { ";",  "ADD" },
    { "_",  "UNPACK" }
};

const unsigned COUNT = sizeof(TABLE) / sizeof(TABLE[0]);

} // namespace

const char * keyword_for_char(uint8_t koi8)
{
    if (!koi8) return 0;
    for (unsigned i = 0; i < COUNT; ++i) {
        std::string keys;
        utf8_to_koi8(TABLE[i].chars, keys);
        for (std::size_t k = 0; k < keys.size(); ++k)
            if (koi8_upper(static_cast<uint8_t>(keys[k])) == koi8)
                return TABLE[i].word;
    }
    return 0;
}

// Зона 5: «в верхнем регистре клавиши этой зоны служат для ввода
// обозначений стандартных математических функций Бейсика» (разд. 2.1).
// На рисунке их восемь: `ARC`, `TAN(`, `SIN(`, `COS(`, `EXP(`, `LOG(`,
// `SQR(`, `ABS(`. Первые четыре знаками не отличить от клавиш зоны 1, а
// байтов у тригонометрии всё равно нет (`docs/format.md`, разд. 5, «Опись»),
// так что клавиш у них здесь нет.
const char * keyword_for_pad(PadFunc kind)
{
    switch (kind) {
    case PAD_MUL: return "ABS(";
    case PAD_ADD: return "SQR(";
    case PAD_SUB: return "EXP(";
    case PAD_DIV: return "LOG(";
    }
    return 0;
}

} // namespace iskra
